カレンダープログラムを作成

masanaito-bita opened the pull request:
動作している様子
https://gyazo.com/e575e64854af2a817e57b2402fec0315

masanaito-bita committed:
カレンダープログラムを作成

masanaito-bita committed:
add package.json

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+const argv = require("minimist")(process.argv.slice(2));
+const date = new Date();
+const year = typeof argv.y !== "undefined" ? argv.y : date.getFullYear();
```
```suggestion
const year = argv.y !== undefined ? argv.y : date.getFullYear();
```

でも良い気がしますー。

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+const month = typeof argv.m !== "undefined" ? argv.m : date.getMonth() + 1;
+
+function cal(year, month) {
+  const first_date = new Date(year, month - 1, 1);
```
JSは`firstDate`のようにキャメルケースを使うのが主流です〜。

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+  const first_date = new Date(year, month - 1, 1);
+  const last_date = new Date(year, month, 0);
+  const end_day_count = last_date.getDate();
+  let week = first_date.getDay();
```
変数の寿命を短くしましょう。

参考 [変数の寿命はできるだけ短くしよう \| FJORD BOOT CAMP（フィヨルドブートキャンプ）](https://bootcamp.fjord.jp/pages/311)

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+  console.log("日 月 火 水 木 金 土");
+
+  let output = "";
+  if (week != 0) {
```
`!==`を使う方がベターです。
ただ、この場合、条件分岐を無くしても同じように動くかも？

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+
+  let output = "";
+  if (week != 0) {
+    process.stdout.write(String(" ").repeat(week * 3));
```
ここだけ`process.stdout.write`を使うより、

```suggestion
    output += String(" ").repeat(week * 3);
```

とした方が出力処理を1まとめにできるのできれいな気がしました。

JunichiIto reviewed (commented):
コメントしましたー。

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -0,0 +1,36 @@
+  }
+
+  for (let i = 1; i <= end_day_count; i++) {
+    if (i < 10) {
+      output += ` ${i} `;
+    } else {
+      output += `${i} `;
+    }
```
下記の関数を使うとシンプルに書けるかも？

https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/String/padStart

masanaito-bita replied:
ありがとうございます。
padStartだと、左端、日曜の行だけスペース(" ")を入れない、という処理が煩雑だったため、`'  ${i}  '`のような記述のままにします。
条件分岐処理はよりシンプルな記述に修正しました。

masanaito-bita committed:
変数名をキャメルケースに修正

masanaito-bita committed:
コード修正

tyoeofからstringで返り値取得を削除
変数定義を使用する直前位置に移動
weekが日曜以外、の条件分岐は不要なので削除
出力を変数outputに渡すよう変更

masanaito-bita committed:
コード修正

三項演算子にする

masanaito-bita committed:
出力の最後に空行を入れる

不要な出力を削除

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -1,36 +1,29 @@
 const date = new Date();
-const year = typeof argv.y !== "undefined" ? argv.y : date.getFullYear();
-const month = typeof argv.m !== "undefined" ? argv.m : date.getMonth() + 1;
+const year = argv.y !== undefined ? argv.y : date.getFullYear();
```
念のため確認してもらいたいのですが、`argv.y`はnumber型になってますか？
string型だと数値計算時に暗黙の型変換が発生するので、場合によっては予期せぬ不具合が出る可能性があります。
yearやmonthには常にnumber型が入るように担保しておく方が良いです。

masanaito-bita replied:
`typeof`で確認したところ、`argv.y`、`argv.m`、`month`、`year`共にnumber型になっていました。
実装の際は暗黙の型変換について意識します！ご確認ありがとうございます。

JunichiIto commented on 02.calendar/cal.js:
```diff
@@ -1,36 +1,29 @@
 
+  const endDayCount = lastDate.getDate();
+  for (let i = 1; i <= endDayCount; i++) {
+    output += i < 10 ? ` ${i} ` : `${i} `;
```
```suggestion
    output += i.toString().padStart(2) + " "
```

と書くこともできそうですが、元のコードも十分シンプルなので、好みの問題になりそうですね。

masanaito-bita replied:
padStartの使い方について理解できました！そのように修正します、ありがとうございます。

JunichiIto reviewed (commented):
コメントしました〜。

masanaito-bita committed:
日にちの位置表示を調整する記述を修正

JunichiIto reviewed (approved):
確認しました。OKです！
